### DISCUSSION TITLE
Why are ".csv files are a scourge upon the earth, and still regrettably common"?

### DISCUSSION THREAD

**Discussion by reagle** [saulpw/visidata#1713] [Q&A] [answered]
Just wondering, how does tsv deal with multi-line items or those with tab characters?

**Comment by saulpw** [comment, marked as answer]
TSV files can't deal with newlines or tab characters in data elements, it's true.  But it is uncommon for a data element to have a newline in it, and extremely rare for (non-binary) data to legitimately contain a tab character.  In both cases, the individual elements can be generally be cleaned to replace the tab or newline with plain whitespace or another replacement character, without corrupting any content.

This one constraint, disallowing row and field separators in data elements, means that .tsv can be extremely simple to parse correctly.  Since all newlines and tabs in TSV files must be row or unit separators, if you come across some corrupt data within the file, you can just wait for the next newline to sync and start parsing correctly again.

This also means you can do other interesting things with TSV files, like stream them in chunks, or  or even split the file into batches of lines and parallelize the parsing.  And you can point to the start of a cell in a TSV file and read the exact string value of the cell; so if you have very long values, reading the data can still be very fast, since you can copy the strings verbatim instead of having to parse them as well.

Compare this with CSV.  Strings quite often have embedded commas, especially outside the US where the comma is the local decimal separator.  There's no reasonable way to replace the comma with another character without changing the data, so CSV demands an escaping or quoting mechanism.  For this, CSV uses the ASCII quote character to (optionally!) wrap string content, but since the ASCII quote character is also fairly common in strings, it needs to be able to escape the quote character itself within a quoted string.  Which it does (in [RFC4180](https://www.rfc-editor.org/rfc/rfc4180) anyway) by using another of the *same quote character*.

Since literal newline characters are allowed in quoted values, a CSV file will be corrupted without chance of recovery if a single errant quote is inserted at the start of any value, since the parser will then be obliged to consider the subsequent text to be part of the value (even the entire rest of the file).  

It didn't have to be this bad; in an alternate universe, CSV might have used a simple local escaping mechanism with an uncommon ASCII escape like C, such that `\n` was a newline and `\c` was a comma and `\\` was a literal backslash.  Now you still have to parse each value, but at least the parsing is straightforward, and there isn't any ambiguity.

But CSV was not designed; it was "created" for some ad hoc purpose and then used and extended over decades without any kind of actual standard (note that RFC4180 above was not written until 2005, by which time it was way too late).   So of course Microsoft Excel, one of the most commonly-used programs to read and write CSV, does something different than the RFC, and now we have several dialects of CSV which are not compatible, and the dialect used is not easily discoverable.

To be fair, TSV files aren't that great either: no data types, no nulls, no multi-line fields, no nested data, no multiple sheets in one file.  So I don't use TSV much myself anymore; I prefer JSONL or Parquet for data interchange and storage.  But sometimes you want something a really simple format and even JSON is too much overhead, so you might reach for a simple text format, and CSV seems like a reasonable candidate because it is ubiquitous and appears (deceptively) simple. But lo! If you just want a simple text format, TSV is about as simple as can be, and will basically work "fine".  Even Microsoft Excel will import it without complaint (which can't be said for "ASV" (ASCII-separated values), which use the proper ASCII separators instead of tab and newline).

But CSV is a well-known nightmare to deal with and should be deprecated as a format immediately and everywhere.   It is a scourge upon the earth.

**Reply by daviewales** [reply]
Ooh. I didn't know about JSONL!

**Comment by reagle** [comment]
Very thorough, thank you!
